drivers/usb/udc/udc_stm32.c

```diff
@@ -0,0 +1,1128 @@
+};
+
+struct udc_stm32_config {
+	uint32_t num_endpoints;
+	uint32_t pma_offset;
+	uint32_t dram_size;
+	uint16_t ep0_mps;
+	uint16_t ep_mps;
+};
+
+static int udc_stm32_lock(const struct device *dev)
+{
+	return udc_lock_internal(dev, K_FOREVER);
+}
+
+static int udc_stm32_unlock(const struct device *dev)
+{
+	return udc_unlock_internal(dev);
+}
+
+#define hpcd2data(hpcd) CONTAINER_OF(hpcd, struct udc_stm32_data, pcd);
+
+void HAL_PCD_ResetCallback(PCD_HandleTypeDef *hpcd)
+{
+	struct udc_stm32_data *priv = hpcd2data(hpcd);
+
+	LOG_DBG("");
+
+	udc_submit_event(priv->dev, UDC_EVT_RESET, 0, NULL);
+}
+
+void HAL_PCD_ConnectCallback(PCD_HandleTypeDef *hpcd)
+{
+	struct udc_stm32_data *priv = hpcd2data(hpcd);
+
+	LOG_DBG("");
+
+	udc_submit_event(priv->dev, UDC_EVT_VBUS_READY, 0, NULL);
+}
+
+void HAL_PCD_DisconnectCallback(PCD_HandleTypeDef *hpcd)
+{
+	struct udc_stm32_data *priv = hpcd2data(hpcd);
+
+	LOG_DBG("");
+
+	udc_submit_event(priv->dev, UDC_EVT_VBUS_REMOVED, 0, NULL);
+}
+
+void HAL_PCD_SuspendCallback(PCD_HandleTypeDef *hpcd)
+{
+	struct udc_stm32_data *priv = hpcd2data(hpcd);
+
+	LOG_DBG("");
+
+	udc_set_suspended(priv->dev, true);
+	udc_submit_event(priv->dev, UDC_EVT_SUSPEND, 0, NULL);
+}
+
+void HAL_PCD_ResumeCallback(PCD_HandleTypeDef *hpcd)
+{
+	struct udc_stm32_data *priv = hpcd2data(hpcd);
+
+	LOG_DBG("");
+
+	udc_set_suspended(priv->dev, false);
+	udc_submit_event(priv->dev, UDC_EVT_RESUME, 0, NULL);
+}
+
+static int usbd_ctrl_feed_dout(const struct device *dev, const size_t length)
+{
+	struct udc_stm32_data *priv = udc_get_private(dev);
+	struct udc_ep_config *cfg = udc_get_ep_cfg(dev, USB_CONTROL_EP_OUT);
+	struct net_buf *buf;
+
+	LOG_DBG("");
+
+	buf = udc_ctrl_alloc(dev, USB_CONTROL_EP_OUT, length);
+	if (buf == NULL) {
+		return -ENOMEM;
+	}
+
+	net_buf_put(&cfg->fifo, buf);
+
+	HAL_PCD_EP_Receive(&priv->pcd, cfg->addr, buf->data, buf->size);
+
+	return 0;
+}
+
+void HAL_PCD_SetupStageCallback(PCD_HandleTypeDef *hpcd)
+{
+	struct udc_stm32_data *priv = hpcd2data(hpcd);
+	struct usb_setup_packet *setup = (void *)priv->pcd.Setup;
+	const struct device *dev = priv->dev;
+	struct net_buf *buf;
+	int err;
+
+	LOG_DBG("");
+
+	buf = udc_ctrl_alloc(dev, USB_CONTROL_EP_OUT,
+			     sizeof(struct usb_setup_packet));
+	if (buf == NULL) {
+		LOG_ERR("Failed to allocate for setup");
+		return;
+	}
+
+	udc_ep_buf_set_setup(buf);
+	memcpy(buf->data, setup, 8);
+	net_buf_add(buf, 8);
+
+	udc_ctrl_update_stage(dev, buf);
+
+	if (!buf->len) {
+		return;
+	}
+
+	if (udc_ctrl_stage_is_data_out(dev)) {
+		/*  Allocate and feed buffer for data OUT stage */
+		err = usbd_ctrl_feed_dout(dev, udc_data_stage_length(buf));
+		if (err == -ENOMEM) {
+			udc_submit_event(dev, UDC_EVT_EP_REQUEST, err, buf);
+		}
+	} else if (udc_ctrl_stage_is_data_in(dev)) {
+		udc_ctrl_submit_s_in_status(dev);
+	} else {
+		udc_ctrl_submit_s_status(dev);
+	}
+
+	if (setup->bRequest == USB_SREQ_SET_ADDRESS) {
+		HAL_PCD_SetAddress(&priv->pcd, setup->wValue);
```

Comment:
This sounds like a race condition. While on modern hosts the time between `udc_ctrl_submit_s_status()` and actually finishing STATUS stage is pretty low, there is no guarantee that STATUS stage will be so fast that it finishes before `HAL_PCD_SetAddress()` is called here.

Reply:
We need to set the Address before the actual submitting of the status stage, I moved that block before the async submit_status to make it more clear. Letting the stack doing it for us after status TX makes the device unresponsive... I know it seems to contradict USB spec, but it is per STM32 UDC spec, so that the UDC is ready to answer host transactions (seems again an internal trick).

Reply:
This behavior deserves a comment in the code.